src/measure_theory/pi.lean

```diff
@@ -252,6 +252,75 @@ lemma pi_hyperplane [∀ i, sigma_finite (μ i)] (i : ι) [has_no_atoms (μ i)]
 show measure.pi μ (eval i ⁻¹' {x}) = 0,
 from pi_eval_preimage_null _ (measure_singleton x)
 
+section intervals
+
+variables {μ} [Π i, partial_order (α i)] [∀ i, sigma_finite (μ i)] [∀ i, has_no_atoms (μ i)]
+
+lemma pi_Iio_ae_eq_pi_Iic {s : set ι} {f : Π i, α i} :
+  pi s (λ i, Iio (f i)) =ᵐ[measure.pi μ] pi s (λ i, Iic (f i)) :=
+begin
+  refine eventually_le.antisymm (eventually_of_forall $ pi_mono (λ i hi, Iio_subset_Iic_self)) _,
+  rw ae_le_set,
+  refine measure_mono_null (pi_diff_pi_subset _ _ _) _,
+  simp only [bUnion_eq_Union, Iic_diff_Iio_same],
+  refine measure_Union_null (λ i, pi_hyperplane _ _ _)
+end
+
+lemma pi_Ioi_ae_eq_pi_Ici {s : set ι} {f : Π i, α i} :
+  pi s (λ i, Ioi (f i)) =ᵐ[measure.pi μ] pi s (λ i, Ici (f i)) :=
+@pi_Iio_ae_eq_pi_Iic ι _ (λ i, order_dual (α i)) _ _ _ _ _ s f
+
+lemma univ_pi_Iio_ae_eq_Iic {f : Π i, α i} :
+  pi univ (λ i, Iio (f i)) =ᵐ[measure.pi μ] Iic f :=
+by { rw ← pi_univ_Iic, exact pi_Iio_ae_eq_pi_Iic }
+
+lemma univ_pi_Ioi_ae_eq_Ici {f : Π i, α i} :
+  pi univ (λ i, Ioi (f i)) =ᵐ[measure.pi μ] Ici f :=
+by { rw ← pi_univ_Ici, exact pi_Ioi_ae_eq_pi_Ici }
+
+lemma pi_Ioo_ae_eq_pi_Icc {s : set ι} {f g : Π i, α i} :
+  pi s (λ i, Ioo (f i) (g i)) =ᵐ[measure.pi μ] pi s (λ i, Icc (f i) (g i)) :=
+begin
+  simp only [← Ioi_inter_Iio, ← Ici_inter_Iic, pi_inter_distrib],
+  exact pi_Ioi_ae_eq_pi_Ici.inter pi_Iio_ae_eq_pi_Iic
+end
+
+lemma univ_pi_Ioo_ae_eq_Icc {f g : Π i, α i} :
+  pi univ (λ i, Ioo (f i) (g i)) =ᵐ[measure.pi μ] Icc f g :=
+by { rw ← pi_univ_Icc, exact pi_Ioo_ae_eq_pi_Icc }
+
+lemma pi_Ioc_ae_eq_pi_Icc {s : set ι} {f g : Π i, α i} :
+  pi s (λ i, Ioc (f i) (g i)) =ᵐ[measure.pi μ] pi s (λ i, Icc (f i) (g i)) :=
+begin
+  simp only [← Ioi_inter_Iic, ← Ici_inter_Iic, pi_inter_distrib],
+  exact pi_Ioi_ae_eq_pi_Ici.inter (eventually_eq.refl _ _)
+end
+
+lemma univ_pi_Ioc_ae_eq_Icc {f g : Π i, α i} :
+  pi univ (λ i, Ioc (f i) (g i)) =ᵐ[measure.pi μ] Icc f g :=
+by { rw ← pi_univ_Icc, exact pi_Ioc_ae_eq_pi_Icc }
+
+lemma pi_Ico_ae_eq_pi_Icc {s : set ι} {f g : Π i, α i} :
+  pi s (λ i, Ico (f i) (g i)) =ᵐ[measure.pi μ] pi s (λ i, Icc (f i) (g i)) :=
+begin
+  simp only [← Ici_inter_Iio, ← Ici_inter_Iic, pi_inter_distrib],
+  exact (eventually_eq.refl _ _).inter pi_Iio_ae_eq_pi_Iic
+end
+
+lemma univ_pi_Ico_ae_eq_Icc {f g : Π i, α i} :
+  pi univ (λ i, Ico (f i) (g i)) =ᵐ[measure.pi μ] Icc f g :=
+by { rw ← pi_univ_Icc, exact pi_Ico_ae_eq_pi_Icc }
+
+end intervals
+
+lemma pi_has_no_atoms (i : ι) [has_no_atoms (μ i)] [∀ i, sigma_finite (μ i)] :
```

Comment:
Is there a reason for this lemma if we have the instance below it?

Reply:
```suggestion
/-- If one of the measures `μ i` has no atoms, them `measure.pi µ`
has no atoms. The instance below assumes that all `μ i` have no atoms. -/
lemma pi_has_no_atoms (i : ι) [has_no_atoms (μ i)] [∀ i, sigma_finite (μ i)] :
```